Handle `fpga seq num` correctly

eschnett opened the pull request.

eschnett committed:
Set `fpga_seq_num` in all stages

eschnett committed:
Set `fpga_seq_num` more consistently

eschnett committed:
Correct `fpga_seq_num` in ringbuffer logic

rhaas80 commented on lib/cuda/cudaCopyToRingbuffer.cpp:
```diff
@@ -151,6 +151,16 @@ cudaEvent_t cudaCopyToRingbuffer::execute(cudaPipelineState& pipestate,
+                               - meta->get_time_downsampling_fpga().at(0)
+                                     * (output_cursor / meta->sample_bytes()));
+        // This should hold most of the time. It will be wrong when we start Kotean with a non-zero
+        // `fpga_seq_num`.
```
we do. This will fail, pretty much every single time I start kotekan on the test node since the synchronization that dpdk does will mean that the very first `fpga_seq_num` passed to kotekan will not be zero (in fact it's rather large for me).

eschnett replied:
Fixed.

geoffryan commented on lib/cuda/cudaCopyFromRingbuffer.cpp:
```diff
@@ -126,9 +126,15 @@ cudaEvent_t cudaCopyFromRingbuffer::execute(cudaPipelineState& pipestate,
     assert(input_cursor % meta->sample_bytes() == 0);
-    meta->set_sample0_offset(meta->get_sample0_offset() + input_cursor / meta->sample_bytes());
+    meta->set_fpga_seq_num(meta->get_fpga_seq_num()
+                           + meta->get_time_downsampling_fpga().at(0)
```
`time_downsampling_fpga` is frequency-dependent, but `fpga_seq_num` is not.  Should the factor here be a new scalar variable like `fpga_seq_stride` or something similar?

eschnett replied:
Yes. I changed this.

geoffryan commented on lib/cuda/cudaCopyToRingbuffer.cpp:
```diff
@@ -151,6 +151,16 @@ cudaEvent_t cudaCopyToRingbuffer::execute(cudaPipelineState& pipestate,
+        // in the beginning only.
+        assert(output_cursor % meta->sample_bytes() == 0);
+        meta->set_fpga_seq_num(meta->get_fpga_seq_num()
+                               - meta->get_time_downsampling_fpga().at(0)
```
same as above

rhaas80 commented on lib/stages/hdf5FileRead.cpp:
```diff
@@ -156,8 +156,12 @@ class hdf5FileRead : public kotekan::Stage {
                 if (dataset.hasAttribute("sample0_offset"))
-                    meta->set_sample0_offset(dataset.getAttribute("sample0_offset").read<int>());
+                    meta->set_sample0_offset(
+                        dataset.getAttribute("sample0_offset").read<std::int64_t>());
```
👍

geoffryan commented on lib/cuda/cudaCorrelator.cpp:
```diff
@@ -120,45 +120,45 @@ cudaEvent_t cudaCorrelator::execute(cudaPipelineState&, const std::vector<cudaEv
-    // TODO: do this automatically in `NDArrayRingBuffer`
+    n2k_corr_meta->set_fpga_seq_num(voltage_meta->get_fpga_seq_num()
+                                    + voltage.get_read_valid().begin()
+                                          * voltage_meta->get_time_downsampling_fpga().at(0));
```
same as above

geoffryan commented on lib/cuda/cudaPL1bitCorrelator.cpp:
```diff
@@ -198,36 +198,33 @@ cudaEvent_t cudaPL1bitCorrelator::execute(cudaPipelineState& /*pipestate*/,
-        div_noremainder(out_meta->get_fpga_seq_num(), n2k_sub_integration_ntime));
+    n2k_counts_meta->set_fpga_seq_num(rfi_meta->get_fpga_seq_num()
+                                      + rfi_RFImask.get_read_valid().begin()
+                                            * rfi_meta->get_time_downsampling_fpga().at(0));
```
same as above

rhaas80 reviewed (changes requested):
Nothing wrong is being added, but a bug persists, namely the assumption that `fpga_seq_num` starts out as 0, which is not true.

geoffryan commented on lib/cuda/cudaCopyFromRingbuffer.cpp:
```diff
@@ -126,9 +126,15 @@ cudaEvent_t cudaCopyFromRingbuffer::execute(cudaPipelineState& pipestate,
+                           + meta->get_time_downsampling_fpga().at(0)
+                                 * (input_cursor / meta->sample_bytes()));
+    meta->set_sample0_offset(meta->get_sample0_offset()
+                             + meta->get_time_downsampling_fpga().at(0)
```
Should this include `offset_downsampling`?

eschnett replied:
That attribute has been removed.

rhaas80 commented on lib/testing/gpuSimulateN2kCorr.cpp:
```diff
@@ -285,6 +285,7 @@ void gpuSimulateN2kCorr::main_thread() {
 
+        // This looks inconsistent
         meta_out->set_fpga_seq_num(meta_in->get_fpga_seq_num());
         meta_out->set_sample0_offset(meta_in->get_sample0_offset() / _sub_integration_ntime);
```
[edited]

It's almost consistent, since: `T_actual = sample0_offset / time_downsampling_fpga` (at least at index T=0) so to get the same `T_actual` when changing `t_downsampling_fpga` by a factor of `_sub_integration_ntime` this needs to be compensated though I would think using a `*` as well if a `*` is used for `time_downsampling_fpga`.

rhaas80 replied:
may resolve iself though: https://chime-experiment.slack.com/archives/CBXE5SMNK/p1766013754753409

geoffryan replied:
Can't see that slack link, but I can say when I wrote that code it was necessary for sample0_offset to explicitly have the downsampling factor, it was treated differently than an fpga tick.  This stage simply replicated what `cudaCorrelator` output, which was governed by the checks in copying data off the GPU.    However, this seems to no longer be the case!  Not sure when that changed.  

Likely a moot point since sample0_offset is being removed :)

eschnett replied:
I went through all (?) the code setting sequence numbers and downsampling factors. I think it's correct now.

rhaas80 commented on lib/cuda/cudaCopyFromRingbuffer.cpp:
```diff
@@ -126,9 +126,15 @@ cudaEvent_t cudaCopyFromRingbuffer::execute(cudaPipelineState& pipestate,
     assert(input_cursor % meta->sample_bytes() == 0);
-    meta->set_sample0_offset(meta->get_sample0_offset() + input_cursor / meta->sample_bytes());
+    meta->set_fpga_seq_num(meta->get_fpga_seq_num()
+                           + meta->get_time_downsampling_fpga().at(0)
```
this is not done quite consistently, sometimes it does `fpga_seq_num = time_downsamplng_fpga * (input_cursor / sample_bytes)` like here, and sometimes it misses the `time_downsampling_fpga` like in https://github.com/kotekan/kotekan/pull/1384/changes#diff-31ced5ba1ee248fe32f1668ff1ffa2fc6abaf6752420acd4bb05754974a71e4dR233

eschnett replied:
I think I fixed all of these.

eschnett committed:
Make time_downsampling_fpga a scalar

Rename metadata attribute time_downsampling_fpga to time_downsampling_fpga_per_frequency.
Introduce a new scalar attribute time_downsampling_fpga.
Use these everywhere.

eschnett committed:
Ensure `fpga_seq_num` is set correctly

eschnett committed:
Remove `sample0_offset` and friends

eschnett committed:
Regenerate kernels

eschnett committed:
Allow non-zero starting `fpga_seq_num`

rhaas80 commented on lib/cuda/cudaCopyNToRingbuffer.cpp:
```diff
@@ -184,8 +181,8 @@ cudaEvent_t cudaCopyNToRingbuffer::execute(cudaPipelineState& /*pipestate*/,
             coarse_freq.at(i) = meta_in->get_coarse_freq()[0];
-            // Check that the sample0 matches for all input buffers
-            assert(meta_ring->get_sample0_offset() == meta_in->get_sample0_offset());
+            // Check that the seq_num and sample0 matches for all input buffers
```
sample0 no longer exists

rhaas80 replied:
fixed in ac8f6413

rhaas80 commented on lib/cuda/cudaFRBBeamReformer.cpp:
```diff
@@ -227,9 +221,10 @@ cudaEvent_t cudaFRBBeamReformer::execute(cudaPipelineState&, const std::vector<c
         assert(input_cursor % in_meta->sample_bytes() == 0);
-        out_meta->set_sample0_offset(div_noremainder(input_cursor, in_meta->sample_bytes()));
+        out_meta->set_fpga_seq_num(in_meta->get_fpga_seq_num()
+                                   + div_noremainder(input_cursor, in_meta->sample_bytes()));
```
this ignores `time_downsampling_fpga` even in the input.

rhaas80 replied:
fixed in d61f67dba

rhaas80 commented on lib/cuda/cudaRFIS012.cpp:
```diff
@@ -189,19 +189,11 @@ cudaEvent_t cudaRFIS012::execute(cudaPipelineState& /*pipestate*/,
-    rfi_S012_meta->set_freq_upchan_factor(freq_upchan_factor);
-    rfi_S012_meta->set_time_downsampling_fpga(time_downsampling_fpga);
+    rfi_S012_meta->set_time_downsampling_fpga(rfi_S012_meta->get_time_downsampling_fpga()
+                                              * rfi_downsampling_factor);
```
this removes the update of ` freq_upchan_factor` is that intentional?

rhaas80 replied:
looks like not changing it is correct, since the RFI computation does not actually change the frequency channels.

rhaas80 commented on lib/cuda/cudaRFIS012bar.cpp:
```diff
@@ -155,19 +155,11 @@ cudaEvent_t cudaRFIS012bar::execute(cudaPipelineState& /*pipestate*/,
-    rfi_S012bar_meta->set_freq_upchan_factor(freq_upchan_factor);
-    rfi_S012bar_meta->set_time_downsampling_fpga(time_downsampling_fpga);
+    rfi_S012bar_meta->set_time_downsampling_fpga(rfi_S012bar_meta->get_time_downsampling_fpga()
+                                                 * rfi_second_downsampling_factor);
```
removes ` freq_upchan_factor`, intentional?

rhaas80 replied:
removal would seem correct since RFI does not change frequency channels.

rhaas80 commented on lib/cuda/cudaRFISKtilde.cpp:
```diff
@@ -191,15 +191,10 @@ cudaEvent_t cudaRFISKtilde::execute(cudaPipelineState& /*pipestate*/,
-    rfi_meta->set_half_fpga_sample0(half_fpga_sample0);
-    rfi_meta->set_time_downsampling_fpga(time_downsampling_fpga);
+    rfi_meta->set_time_downsampling_fpga(rfi_meta->get_time_downsampling_fpga()
+                                         * div_noremainder(128 * 8, rfi_downsampling_factor));
```
this seems odd. Is it really the case that fro `rfi_downsampling_factor` of 1024 there would not be any time downsampling in this step (and there would be maximum downsampling for `rfi_downsampling_factor` of 1)?

rhaas80 replied:
looks like it, yes, since rfimask undoes `rfi_downsampling-factor`

rhaas80 commented on lib/stages/lostSamplesToPLMask.cpp:
```diff
@@ -162,25 +162,15 @@ void lostSamplesToPLMask::main_thread() {
-                    PL_MASK_FREQS_PER_BIN, PL_MASK_DOWNSAMPLING_FACTOR * PL_MASK_HILO_SPLIT);
-                pl_mask_meta->set_time_downsampling_fpga(time_downsampling_fpga);
+                pl_mask_meta->set_time_downsampling_fpga(PL_MASK_DOWNSAMPLING_FACTOR
+                                                         * PL_MASK_HILO_SPLIT);
```
does not take possible downsampling of lostSamples into account (which is always 1 in the current dpdk code)

rhaas80 replied:
fixed in 529245476

rhaas80 commented on lib/stages/lostSamplesToPLMask.cpp:
```diff
@@ -197,21 +187,8 @@ void lostSamplesToPLMask::main_thread() {
-                                                      time_downsampling_fpga.end());
-                pl_mask_meta->set_time_downsampling_fpga(pl_mask_time_downsampling_fpga);
+                pl_mask_meta->set_time_downsampling_fpga(PL_MASK_DOWNSAMPLING_FACTOR
+                                                         * PL_MASK_HILO_SPLIT / 2);
```
wrong value, sets the same scalar twice.

rhaas80 replied:
fixed in 529245476

rhaas80 commented on lib/testing/testN2kGen.cpp:
```diff
@@ -240,26 +233,20 @@ void testN2kGen::set_counts_metadata(const std::shared_ptr<chordMetadata>& meta,
     meta->set_array_dimension(4, count_blocksize, "D8Plo2");
     meta->set_strides_simple();
 
+    // This looks inconsistent
```
does it still do so?

rhaas80 reviewed (changes requested):
one actual bug I think, setting `time_downsampling_fpga` in `lostSamplesToPLMask`

rhaas80 commented:
Additional changes in https://github.com/kotekan/kotekan/pull/1386

eschnett commented:
Superseded by https://github.com/kotekan/kotekan/pull/1386
